scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -4361,37 +4369,46 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
         token.isNot[LeftBrace] && !isColonEol(token)
       }
     }
-    def bracelessPackageStats(): List[Stat] = token match {
-      case _: EOF => Nil
+    val buf = new ListBuffer[Stat]
+    @tailrec
+    def bracelessPackageStats(f: List[Stat] => List[Stat]): List[Stat] = token match {
+      case _: EOF => f(buf.toList)
       case StatSep() =>
         next()
-        bracelessPackageStats()
+        bracelessPackageStats(f)
       case _: KwPackage if !ahead(token.is[KwObject]) =>
         val startPos = in.tokenPos
         next()
         val qid = qualId()
         def inPackage(stats: => List[Stat]) = {
-          val pkg = autoEndPos(startPos)(Pkg(qid, stats))
+          buf += autoEndPos(startPos)(Pkg(qid, stats))
           acceptStatSepOpt()
-          pkg +: bracelessPackageStats()
         }
         token match {
-          case _: LeftBrace => inPackage(inBraces(statSeq(statpf)))
+          case _: LeftBrace =>
+            inPackage(inBraces(statSeq(statpf)))
+            bracelessPackageStats(f)
           case t if isColonEol(t) =>
             next()
             in.observeIndented()
             inPackage(indented(statSeq(statpf)))
-          case _ => List(autoEndPos(startPos)(Pkg(qid, bracelessPackageStats())))
+            bracelessPackageStats(f)
+          case _ => bracelessPackageStats(x => f(List(autoEndPos(startPos)(Pkg(qid, x)))))
```

Comment:
it's not identity here

Reply:
I haven't noticed this one. Wouldn't it be simpler to just  do:
```
val stats = bracelessPackageStats()
List(autoEndPos(startPos)(Pkg(qid, stats))
```
to avoid passing the f function?

Reply:
then you can't do `@tailrec` :)

Reply:
ach, that's because of the tail recursion.